src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/MessagePackHubProtocol.cs

```diff
@@ -0,0 +1,237 @@
+            {
+                throw new FormatException("Message type is missing.");
+            }
+
+            switch (messageType)
+            {
+                case InvocationMessageType:
+                    return CreateInvocationMessage(unpacker, binder);
+                case StreamItemMessageType:
+                    return CreateStreamItemMessage(unpacker, binder);
+                case CompletionMessageType:
+                    return CreateCompletionMessage(unpacker, binder);
+                default:
+                    throw new FormatException($"Invalid message type: {messageType}.");
+            }
+        }
+
+        private static InvocationMessage CreateInvocationMessage(Unpacker unpacker, IInvocationBinder binder)
+        {
+            var invocationId = ReadInvocationId(unpacker);
+            var nonBlocking = ReadBoolean(unpacker, "nonBlocking");
+            var target = ReadString(unpacker, "target");
+            var argumentCount = ReadArrayLength(unpacker, "arguments");
+            var parameterTypes = binder.GetParameterTypes(target);
+
+            if (parameterTypes.Length != argumentCount)
+            {
+                throw new FormatException(
+                    $"Target method expects {parameterTypes.Length} arguments(s) but invocation has {argumentCount} argument(s).");
+            }
+
+            var arguments = new object[argumentCount];
+            for (var i = 0; i < argumentCount; i++)
+            {
+                arguments[i] = DeserializeObject(unpacker, parameterTypes[i], "argument");
+            }
+
+            return new InvocationMessage(invocationId, nonBlocking, target, arguments);
+        }
+
+        private static StreamItemMessage CreateStreamItemMessage(Unpacker unpacker, IInvocationBinder binder)
+        {
+            var invocationId = ReadInvocationId(unpacker);
+            var itemType = binder.GetReturnType(invocationId);
+            var value = DeserializeObject(unpacker, itemType, "item");
+            return new StreamItemMessage(invocationId, value);
+        }
+
+        private static CompletionMessage CreateCompletionMessage(Unpacker unpacker, IInvocationBinder binder)
+        {
+            var invocationId = ReadInvocationId(unpacker);
+            var error = ReadString(unpacker, "error");
+
+            var hasResult = false;
+            object result = null;
+            if (error == null)
+            {
+                hasResult = ReadBoolean(unpacker, "hasResult");
+                if (hasResult)
+                {
+                    var itemType = binder.GetReturnType(invocationId);
+                    result = DeserializeObject(unpacker, itemType, "argument");
+                }
+            }
+
+            return new CompletionMessage(invocationId, error, result, hasResult);
+        }
+
+        // TODO: when to return false?
+        public bool TryWriteMessage(HubMessage message, Stream output)
```

Comment:
They could be merged indeed. I am planning to review the Try* methods we have since they oftentimes never return false (anymore).